os/pm/pm_timedStay.c

```diff
@@ -0,0 +1,171 @@
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/************************************************************************
+ * pm/pm_timedStay.c
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name NuttX nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/wdog.h>
+
+#include <pm_timer/pm_timer.h>
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Private Variables
+ ************************************************************************/
+
+/* This array maps the pid to their respective wdog timer. 
+ * Assumption: when a thread does a timed lock , only the same thread 
+ * can unlock the "timed lock" before it expire */
+static WDOG_ID pid_timerMap[CONFIG_MAX_TASKS];
```

Comment:
This may cause potential operational errors in the unloading of the binary or in the terminating situation of the task.

Isn't it necessary to handle this part using APIs such as atexit() and on_exit()?

Reply:
The pid can be over than CONFIG_MAX_TASKS.

#define MAX_PID_MASK	(CONFIG_MAX_TASKS - 1)
#define PIDHASH(pid)	((pid) & MAX_PID_MASK)
task_assignpid()

Please ref this logic

Reply:
Let's seperate PR about this and pm_timedStay_cancel.

we can apply pm_timedStay_cancel later.
This PR being delayed.

For fast merge, Please remove it in this PR and please raise new PR.

Reply:
done